### PR TITLE
Unnest ERBProcessor to keep Sprockets out the constant scope chain

### LINKED ISSUE #394: Rails.application no longer available in sprockets assets
### Expected behavior

I have an existing `foo.coffee.erb` file that does this in sprockets 3:

```coffee
constants = {
  SOME_VAR:       '<%= Rails.application.secrets.some_var %>'
}
```

### Actual behavior

On master sprockets, I get: `undefined method 'application' for Sprockets::Rails:Module`. This is after many attempts to diagnose an even more cryptic `NoMethodError - undefined method 'env' for Sprockets::Rails:Module:`. I did not get the "real" error message until I deleted all my javascript and started adding it back line by line.

### System configuration

- Sprockets version: master
- Ruby version: 2.3.1

### Example App

n/a atm

### PR DESCRIPTION
**Pull request by matthewd**
cc @schneems @bouk 

Reverts #396, restoring #365; fixes #394

### CODE REVIEW

File: lib/sprockets/erb_processor.rb [hunk 1]
```diff
@@ -1,31 +1,32 @@
...
 
-module Sprockets
-  class ERBProcessor
-    # Public: Return singleton instance with default options.
-    #
-    # Returns ERBProcessor object.
-    def self.instance
-      @instance ||= new
-    end
+class Sprockets::ERBProcessor
```
Comment (bouk) [inline lib/sprockets/erb_processor.rb]: Clever!

You might need to add a 'module Sprockets;end' before this though, in case just this file is imported

Comment (matthewd) [inline lib/sprockets/erb_processor.rb]: I was going to, but it gets loaded from `sprockets.rb`, inside (and thus after) the main `module Sprockets`.

As this file doesn't strike me as one people are likely to be requiring directly, it didn't seem worth bothering.

**Comment by bouk** [comment]
lgtm

**Comment by schneems** [comment]
First of all, awesome job. Second. This is crazy. I had no idea that nesting a class creation would yield different code. Played around with https://www.diffchecker.com/hhOcwU6P.

```
CONSTANT = "main scope"

module Sprockets
  CONSTANT = "sprockets scope"
end

module Sprockets
  class NestedScope
    puts CONSTANT
  end
end
# => sprockets scope

class Sprockets::NonNestedScope
  puts CONSTANT
end
# => main scope
```
I kinda think I understand what's going on. It looks like when Ruby encounters a constant it walks up the scope chain looking for the first instance of that constant, and defining the class like `Sprockets::NonNestedScope` does not create a `Sprockets` scope. so it fixes our problem. I'm wondering if that is intended behavior or a weird implementation side effect. I would have assumed that the two class declarations were doing the same thing, but apparently they're not.

This is legitimately the most surprising thing i've learned about Ruby, perhaps in the last year. Great job, would love to hear more about the context of how you found this behavior or figured out how to pair it with this solution.

**Comment by bouk** [comment]
I was having some trouble fully understanding what was going on, so I made this example to explain it for myself https://eval.in/659559

The behavior is definitely confusing, especially the second case where the inner class constant takes preference over the higher-level constant, but not if it's called inside the module namespace

**Comment by schneems** [comment]
Thanks to both of you for this feature and the fix.

**Comment by bouk** [comment]
I asked @charliesome to shine some light on how constant lookup works and what is going on here, so read and learn!

> constant scope is ruby is *mostly* static
> the way it’s implemented is there’s a CREF stack
> CREF being constant ref
> which is a linked list
> certain constructs in the language push a new node to the cref stack, like class, module, etc.
> when you reference a constant it compiles to the getconstant instruction: https://github.com/ruby/ruby/blob/trunk/insns.def#L180-L199
> when you reference a constant it compiles to the getconstant instruction: https://github.com/ruby/ruby/blob/trunk/insns.def#L180-L199
> that calls vm_get_ev_const (https://github.com/ruby/ruby/blob/trunk/vm_insnhelper.c#L758) which walks up the cref stack attempting to look up the constant in each class on the cref stack
> each individual scope lookup walks the class’s ancestry too
> so it’s sort of like a two-dimensional lookup
> when you call instance_eval it grabs the singleton class of the object you’re instance_eval’ing in: https://github.com/ruby/ruby/blob/trunk/vm_eval.c#L1707-L1712
> when you call instance_eval it grabs the singleton class of the object you’re instance_eval’ing in: https://github.com/ruby/ruby/blob/trunk/vm_eval.c#L1707-L1712
> and it passes that to specific_eval
> specific_eval calls eval_under which pushes the klass (in this case, the singleton class of the instance_eval receiver) to the cref stack: https://github.com/ruby/ruby/blob/trunk/vm_eval.c#L1630
> specific_eval calls eval_under which pushes the klass (in this case, the singleton class of the instance_eval receiver) to the cref stack: https://github.com/ruby/ruby/blob/trunk/vm_eval.c#L1630
> oh! found out why that weird behaviour is happening
> so I compiled ruby with some debug logging: https://gist.github.com/charliesome/9ca4989828f92bb0d80d6dcf4c07ef92
> so I compiled ruby with some debug logging: https://gist.github.com/charliesome/9ca4989828f92bb0d80d6dcf4c07ef92
> you can see in the A::B section it fails to find the constant while walking up the cref stack
> except when it prints “nested scope”, which is where it finds CONSTANT in A while walking the cref stack
> but at that point, the singleton class of the object doesn’t have any constant set in it so it’s skipped and then it hits A
> however another detail of vm_get_ev_const is that if the cref walk fails to find the const it’ll look up the constant in self: https://github.com/ruby/ruby/blob/trunk/vm_insnhelper.c#L809-L822
> however another detail of vm_get_ev_const is that if the cref walk fails to find the const it’ll look up the constant in self: https://github.com/ruby/ruby/blob/trunk/vm_insnhelper.c#L809-L822
> (I made a mistake before when I said the cref walk also checks ancestry - it doesn’t. the cref walk calls rb_const_lookup which just checks the constant table of the class pointed to by each cref node. the self lookup calls rb_const_get which searches the ancestry of the class you’re getting the const from)
> pretty fun edge case! it had me digging 😛

**Comment by francisco-rojas** [comment]
@schneems @bouk 
I'm trying to get this to work but I'm not sure how this feature must be used. 
Simply using ENV in .sass.erb files doesn't work. I also tried other approaches to no avail.

If you can provide some guidance that would be much appreciated as this is not documented anywhere I searched. Maybe I'm using a version of rails that doesn't implement this feature? I am trying rails 4.2.6 and 4.2.8

**Comment by schneems** [comment]
@francisco-rojas can you give us an example app where this doesn't work. Unless sass provides it's own ERB rendering (which I don't think it does) this should work out of the box ™️.

You should use an env var in your erb

```
<%= ENV["WELCOME_TEXT"] %>
```

When you compile

```
$ RAILS_ENV=production WELCOME_TEXT=hi bundle exec rake assets:precompile
```

It should generate with your text. Then when you change the text, it should re-compile the file

```
$ RAILS_ENV=production WELCOME_TEXT=hello bundle exec rake assets:precompile
```

**Comment by francisco-rojas** [comment]
@schneems 

Here: https://github.com/francisco-rojas/assets_env
That's using rails 4.2.6 (which I'm not sure already includes this fix). 

I noticed a couple of things:

1. Quoting the erb code like this `color: "<%= ENV['H1_COLOR'] %>"`  behaves differently from the unquoted erb `color: <%= ENV['H1_COLOR'] %>`. When unquoted it changes the color if I add some changes to the file (even if I just add new lines and restart the server with the new H1_COLOR value) like this: `H1_COLOR=red rails s` . When quoted it does not change the color.
See: https://github.com/francisco-rojas/assets_env/blob/master/app/assets/stylesheets/home.sass.erb
2. If I use the unquoted erb, stop the server and restart it with a new value for H1_COLOR it does not change the color, until, as mentioned above, I add some changes to the file so rails knows that it needs to precompile

Let me know if you want me to try a different version of rails.

**Pull request merged**